File: api/src/db/models/lab/allocatable/lab_allocation.py
from abc import abstractmethod
from datetime import date, datetime, timezone
from typing import ClassVar, Generic, Self, Type, TypeVar
from uuid import UUID, uuid4

from sqlalchemy import Column, ForeignKey, Table
from sqlalchemy.orm import Mapped, mapped_column, relationship, validates
from sqlalchemy.dialects import postgresql

from db import local_object_session
from db.models.fields import uuid_pk
from db.models.base.base import Base, model_id
from db.models.lab.lab import Lab
from db.models.user import User
from db.models.lab.provisionable import LabProvision

from .allocatable import Allocatable
from .allocation_consumer import LabAllocationConsumer
from .allocation_status import (
    ALLOCATION_STATUS_TRANSITION,
    AllocationStatus,
    ALLOCATION_STATUS_ENUM,
    AllocationStatusError,
    AllocationStatusTransition,
)


TAllocatable = TypeVar("TAllocatable", bound=Allocatable)


lab_allocation_setup_provisions = Table(
    "lab_allocation_setup_provisions",
    Base.metadata,
    Column("allocation_id", ForeignKey("lab_allocation.id"), primary_key=True),
    Column("provision_id", ForeignKey("lab_provision.id"), primary_key=True),
)

lab_allocation_teardown_provisions = Table(
    "lab_allocation_teardown_provisions",
    Base.metadata,
    Column("allocation_id", ForeignKey("lab_allocation.id"), primary_key=True),
    Column("provision_id", ForeignKey("lab_provision.id"), primary_key=True),
)


class LabAllocation(Base, Generic[TAllocatable]):
    __allocation_type__: ClassVar[str]
    __tablename__ = "lab_allocation"

    def __init_subclass__(cls, **kw) -> None:
        if not hasattr(cls, "__allocation_type__"):
            raise TypeError("LabAllocation subclass haa no __allocation_type__")
        return super().__init_subclass__(**kw)

    id: Mapped[uuid_pk] = mapped_column()
    type: Mapped[str] = mapped_column(postgresql.VARCHAR(64))

    @validates("type")
    def validate_type(self, value, key):
        if self.type:
            raise ValueError("type cannot be modified after creation")
        return value

    def __init__(self, lab: Lab, requested_by: User, requested_note: str):
        self.type = self.__allocation_type__
        self.id = uuid4()
        self.lab_id = lab.id
        self._set_request(requested_by, note=requested_note)
        super().__init__()

    status: Mapped[AllocationStatus] = mapped_column(
        ALLOCATION_STATUS_ENUM, default=AllocationStatus.REQUESTED
    )

    lab_id: Mapped[UUID] = mapped_column(ForeignKey("lab.id"))
    lab: Mapped[Lab] = relationship()

    consumer_type: Mapped[str] = mapped_column(postgresql.VARCHAR(64))
    consumer_id: Mapped[UUID] = mapped_column(ForeignKey("lab_allocation_consumer.id"))
    consumer: Mapped[LabAllocationConsumer] = relationship()

    start_date: Mapped[date | None] = mapped_column(postgresql.DATE, nullable=True)
    end_date: Mapped[date | None] = mapped_column(postgresql.DATE, nullable=True)


    @abstractmethod
    async def get_target(self):
        raise NotImplementedError

    request_at: Mapped[datetime] = mapped_column(postgresql.TIMESTAMP(timezone=True))
    request_by_id: Mapped[UUID] = mapped_column(ForeignKey("user.id"))
    request_by: Mapped[User] = relationship()
    request_note: Mapped[str] = mapped_column(postgresql.TEXT)

    @property
    def request(self):
        return AllocationStatusTransition(
            allocation_id=self.id,
            status=AllocationStatus.REQUESTED,
            at=self.request_at,
            by_id=self.request_by_id,
            note=self.request_note,
        )

    previous_requests: Mapped[list[AllocationStatusTransition]] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.REQUESTED, repeatable=True),
        server_default="[]",
    )

    @property
    def requests(self):
        return [self.request, *self.previous_requests]

    def _set_request(self, request_by: User, *, note: str = ""):
        self.previous_requests = [self.request, *self.previous_requests]

        self.request_at = datetime.now(tz=timezone.utc)
        self.request_by_id = request_by.id
        self.request_note = note

    @property
    def last_requested(self):
        return self.last_requested[-1]

    async def resubmit(self, by: User, note: str) -> Self:
        # TODO: check can be submitted by the user

        if self.status not in {AllocationStatus.REJECTED, AllocationStatus.APPROVED}:
            raise AllocationStatusError(
                self.status,
                AllocationStatus.REQUESTED,
                "Can only resubmit a rejected or approved error",
            )
        self._set_request(by, note=note)
        return await self.save()

    approval: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.APPROVED), nullable=True
    )

    @property
    def is_approved(self):
        return self.approval is not None

    @property
    def approved_by_id(self):
        return self.approval['by_id'] if self.approval else None

    @property
    def approved_at(self):
        return self.approval['at'] if self.approval else None

    async def approve(self, by: User, note: str):
        if self.status not in {AllocationStatus.REQUESTED, AllocationStatus.REJECTED}:
            raise AllocationStatusError(
                self.status,
                AllocationStatus.APPROVED,
                "can only approve a requested or rejected allocation",
            )

        self._approved = self.__allocation_status_metadata(
            AllocationStatus.APPROVED, by, note
        )

    rejections: Mapped[list[AllocationStatusTransition]] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.REJECTED, repeatable=True),
        server_default="[]",
    )

    @property
    def rejection(self):
        return self.rejections[-1] if self.rejections else None

    @property
    def is_rejected(self):
        return self.rejection is not None

    @property
    def rejected_by_id(self):
        return self.rejection['by_id'] if self.rejection else None

    @property
    def rejected_at(self):
        return self.rejection['at'] if self.rejection else None

    async def reject(self, by: User, note: str) -> Self:
        if self.status not in {AllocationStatus.REQUESTED, AllocationStatus.REJECTED}:
            raise AllocationStatusError(
                self.status,
                AllocationStatus.REJECTED,
                "can only reject a requested allocation",
            )

        self.rejections = [
            *self.rejections,
            self.__allocation_status_metadata(AllocationStatus.REJECTED, by, note),
        ]
        return await self.save()

    denial: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.DENIED), nullable=True
    )

    @property
    def is_denied(self):
        return self.denial is not None

    @property
    def denied_by_id(self):
        return self.denial['by_id'] if self.denial else None

    @property
    def denied_at(self):
        return self.denial['at'] if self.denial else None

    async def deny(self, by: User, note: str) -> Self:
        """
        A denial is a strong rejection, request is closed and cannot be
        resubmitted for editing
        """

        if self.status != AllocationStatus.REQUESTED:
            raise AllocationStatusError(
                self.status,
                AllocationStatus.DENIED,
                f"Can only deny a requested allocation",
            )

        self.denial = self.__allocation_status_metadata(
            AllocationStatus.REQUESTED, by, note
        )
        return await self.save()

    setup: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.SETUP), nullable=True
    )

    @property
    def setup_begin_at(self):
        return self.setup['at'] if self.setup else None

    @property
    def setup_by_id(self):
        return self.setup['by_id'] if self.setup else None

    setup_provisions: Mapped[list[LabProvision]] = relationship(
        secondary=lab_allocation_setup_provisions
    )

    async def has_pending_setup_provisions(self) -> bool:
        raise NotImplementedError

    async def begin_setup(self, by: User, note: str) -> Self:
        if self.status != AllocationStatus.APPROVED:
            raise AllocationStatusError(
                self.status,
                AllocationStatus.SETUP,
                f"Can only begin setup of an approved allocation",
            )

        self._setup = self.__allocation_status_metadata(
            AllocationStatus.SETUP, by, note
        )
        return await self.save()

    prepared: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.PREPARED), nullable=True
    )

    @property
    def prepared_at(self):
        return self.prepared['at'] if self.prepared else None

    async def mark_as_prepared(self, note: str):
        if self.status != AllocationStatus.SETUP:
            raise AllocationStatusError(self.status, AllocationStatus.PREPARED, f"")

        setup_by_id = self.setup['by_id']

        self.prepared = self.__allocation_status_metadata(
            AllocationStatus.PREPARED, setup_by_id, note
        )

    progress_events: Mapped[list[AllocationStatusTransition]] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.IN_PROGRESS, repeatable=True),
        server_default="[]",
    )

    @property
    def commenced_event(self):
        return self.progress_events[0] if self.progress_events else None

    @property
    def is_commenced(self):
        return bool(self.progress_events)

    @property
    def commenced_at(self):
        return self.commenced_event['at'] if self.progress_events else None

    @property
    def commenced_by_id(self):
        return self.commenced_event['by_id'] if self.progress_events else None

    async def begin(self, by: User, note: str) -> Self:
        if self.status != AllocationStatus.PREPARED:
            raise AllocationStatusError(
                self.status, AllocationStatus.IN_PROGRESS, f"unprepared allocation"
            )
        self.in_progress = [
            self.__allocation_status_metadata(AllocationStatus.IN_PROGRESS, by, note)
        ]
        return await self.save()

    async def add_progress(self, note: str) -> Self:
        if self.status != AllocationStatus.IN_PROGRESS:
            raise AllocationStatusError(
                self.status, AllocationStatus.IN_PROGRESS, f"Allocation not in progress"
            )

        self.in_progress = [
            *self.in_progress,
            self.__allocation_status_metadata(AllocationStatus.IN_PROGRESS, self.commenced_by_id, note),
        ]
        return await self.save()

    completion: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.COMPLETED), nullable=True
    )

    @property
    def is_completed(self):
        return self.completion is not None

    @property
    def completed_at(self):
        return self.completion['at'] if self.completion else None

    @property
    def completed_by_id(self):
        return self.completion['by_id'] if self.completion else None

    async def complete(self, by: User):
        if self.status != AllocationStatus.IN_PROGRESS:
            raise AllocationStatusError(
                self.status, AllocationStatus.IN_PROGRESS, f"allocation not in progress"
            )

        self.completed = self.__allocation_status_metadata(
            AllocationStatus.COMPLETED, by, note="completed"
        )

        self = await self.save()

        requires_teardown = await self.has_pending_teardown_provisions()
        if not requires_teardown:
            return await self.finalise(by, note="no teardown required")

        return self

    cancellation: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.CANCELLED), nullable=True
    )

    @property
    def is_cancelled(self):
        return self.cancellation is not None

    @property
    def cancelled_at(self):
        return self.cancellation['at'] if self.cancellation else None

    @property
    def cancelled_by_id(self):
        return self.cancellation['by_id'] if self.cancellation else None

    async def cancel(self, by: User, note: str) -> Self:
        if not self.status.is_cancellable:
            raise AllocationStatusError(
                self.status,
                AllocationStatus.CANCELLED,
                "Source status is not cancellable",
            )

        self.cancelled = self.__allocation_status_metadata(
            AllocationStatus.CANCELLED, by, note
        )

        await self.begin_teardown(by, note="from_cancellation")
        requires_teardown = self.has_pending_teardown_provisions()

        return await self.save()

    teardown_provisions: Mapped[list[LabProvision]] = relationship(
        secondary=lab_allocation_teardown_provisions
    )

    teardown: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.TEARDOWN), nullable=True
    )

    @property
    def teardown_begin_at(self):
        return self.teardown['at'] if self.teardown else None

    @property
    def teardown_by_id(self):
        return self.teardown['by_id'] if self.teardown else None


    async def begin_teardown(self, triggered_by: User, note: str) -> Self:
        teardown_provisions: list[LabProvision] = (
            await self.awaitable_attrs.teardown_provisions
        )

        for provision in teardown_provisions:
            work = await provision.get_or_create_work()

        return self

    async def has_pending_teardown_provisions(self) -> bool:
        raise NotImplementedError

    finalisation: Mapped[AllocationStatusTransition | None] = mapped_column(
        ALLOCATION_STATUS_TRANSITION(AllocationStatus.FINALISED), nullable=True
    )

    @property
    def is_finalised(self):
        return self.finalisation is not None

    @property
    def finalised_at(self):
        return self.finalisation['at'] if self.finalisation else None

    @property
    def finalised_by_id(self):
        return self.finalisation['by_id'] if self.finalisation else None

    async def mark_as_final(self, *, note: str) -> Self:
        if self.status not in {
            AllocationStatus.COMPLETED,
            AllocationStatus.CANCELLED,
        }:
            raise AllocationStatusError(
                self.status,
                AllocationStatus.FINALISED,
                "must be completed or cancelled",
            )

        requires_teardown = await self.has_pending_teardown_provisions()
        if requires_teardown:
            raise AllocationStatusError(
                self.status, AllocationStatus.FINALISED, "pending teardown tasks"
            )

        self.finalisation = self.__allocation_status_metadata(
            AllocationStatus.FINALISED, self.teardown['by_id'], note
        )
        return await self.save()

    def __allocation_status_metadata(
        self, status: AllocationStatus, by: User | UUID, note: str
    ):
        return AllocationStatusTransition(
            allocation_id=self.id,
            status=status,
            at=datetime.now(tz=timezone.utc),
            by_id=model_id(by),
            note=note,
        )

File: app/src/app/lab/common/allocatable/lab-allocation.ts
import { inject } from "@angular/core";
import { FormBuilder, FormGroup } from "@angular/forms";
import { formatISO, parseISO } from "date-fns";
import { Model, ModelCreateRequest, ModelIndexPage } from "src/app/common/model/model";
import { isJsonObject, JsonObject } from "src/app/utils/is-json-object";
import { isUUID } from "src/app/utils/is-uuid";

export interface Allocatable<TAllocation extends LabAllocation<any>> {
    allocationType: string;
    activeAllocations: ModelIndexPage<TAllocation>;
}

export const ALLOCATION_STATUSES = [
    'requested',
    'approved',
    'rejected',
    'denied',
    'setup',
    'prepared',
    'in_progress',
    'completed',
    'cancelled',
    'teardown',
    'finalised'
] as const;
export type AllocationStatus = typeof ALLOCATION_STATUSES[number];
export function isAllocationStatus(obj: unknown): obj is AllocationStatus {
    return typeof obj === 'string' && ALLOCATION_STATUSES.includes(obj as any);
}

export interface AllocationEvent {
    status: AllocationStatus;
    at: Date;
    byId: string;
}

function allocationEventFromJson(json: JsonObject): AllocationEvent {
    if (!isAllocationStatus(json['status'])) {
        throw new Error("Expected an allocation status 'status'");
    }
    if (typeof json['at'] !== 'string') {
        throw new Error("Expected a datetime string 'at'");
    }
    if (!isUUID(json['byId'])) {
        throw new Error("Expected a UUID 'byId'");
    }
    return {
        status: json['status'],
        at: parseISO(json['at']),
        byId: json['byId']
    };
}

export class LabAllocation<TAllocatable extends Allocatable<any>> extends Model {
    readonly type: string;
    readonly labId: string;
    readonly status: AllocationStatus

    readonly consumerType: string;
    readonly consumerId: string;

    readonly startDate: Date | null;
    readonly endDate: Date | null;

    readonly requestAt: Date;
    readonly requestById: string;

    readonly allRequests: ReadonlyArray<AllocationEvent>;

    readonly isApproved: boolean;
    readonly approvedAt: Date | null;
    readonly approvedById: string | null;

    readonly isRejected: boolean;
    readonly rejectedAt: Date | null;
    readonly rejectedById: string | null;

    readonly allRejections: ReadonlyArray<AllocationEvent>;

    readonly isDenied: boolean;
    readonly deniedAt: Date | null;
    readonly deniedById: string | null;

    readonly setupBeginAt: Date | null;
    readonly setupById: string | null;

    readonly isPrepared: boolean;
    readonly preparedAt: Date | null;

    readonly isCommenced: boolean;
    readonly commencedAt: Date | null;
    readonly commencedById: string | null;

    readonly progressEvents: ReadonlyArray<AllocationEvent>;

    readonly isCompleted: boolean;
    readonly completedAt: Date | null;
    readonly completedById: string | null;

    readonly isCancelled: boolean;
    readonly cancelledAt: Date | null;
    readonly cancelledById: string | null;

    readonly teardownBeginAt: Date | null;
    readonly teardownById: string | null;

    readonly isFinalised: boolean;
    readonly finalisedAt: Date | null;
    readonly finalisedById: string | null;


    constructor(json: JsonObject) {
        super(json);

        if (typeof json['type'] !== 'string') {
            throw new Error("Expected a string 'type'");
        }
        this.type = json['type'];

        if (!isUUID(json['labId'])) {
            throw new Error("Expected a UUID 'labId'");
        }
        this.labId = json['labId'];

        if (typeof json['consumerType'] !== 'string') {
            throw new Error(`Expected a string 'consumerType'`);
        }
        this.consumerType = json['consumerType'];

        if (!isUUID(json['consumerId'])) {
            throw new Error("Expected a uuid 'consumerId'");
        }
        this.consumerId = json['consumerId'];

        if (!isAllocationStatus(json['status'])) {
            throw new Error("Expected an allocation status 'status'");
        }
        this.status = json['status'];

        if (typeof json['startDate'] == null) {
            this.startDate = null;
        } else if (typeof json['startDate'] === 'string') {
            this.startDate = parseISO(json['startDate']);
        } else {
            throw new Error(`Expected a datetime string or null 'startDate'`);
        }

        if (typeof json['endDate'] == null) {
            this.endDate = null;
        } else if (typeof json['endDate'] === 'string') {
            this.endDate = parseISO(json['endDate']);
        } else {
            throw new Error(`Expected a datetime string or null 'endDate'`);
        }


        if (typeof json['requestAt'] !== 'string') {
            throw new Error("Expected a datetime string 'requestAt'");
        }
        this.requestAt = parseISO(json['requestAt']);

        if (!isUUID(json['requestById'])) {
            throw new Error("Expected a UUID 'requestById'");
        }
        this.requestById = json['requestById'];

        if (!Array.isArray(json['allRequests']) || !json['allRequests'].every(isJsonObject)) {
            throw new Error("Expected an array of json objects 'allRequests'");
        }
        this.allRequests = json['allRequests'].map(allocationEventFromJson);

        if (typeof json['isApproved'] !== 'boolean') {
            throw new Error("Expected a boolean 'isApproved'");
        }
        this.isApproved = json['isApproved'];

        if (json['approvedAt'] == null) {
            this.approvedAt = null;
        } else if (typeof json['approvedAt'] === 'string') {
            this.approvedAt = parseISO(json['approvedAt']);
        } else {
            throw new Error("Expected a datetime string 'approvedAt'")
        }

        if (json['approvedById'] == null) {
            this.approvedById = null;
        } else if (isUUID(json['approvedById'])) {
            this.approvedById = json['approvedById']
        } else {
            throw new Error("Expected a datetime string 'approvedAt'")
        }

        if (typeof json['isRejected'] !== 'boolean') {
            throw new Error("Expected a boolean 'isRejected'");
        }
        this.isRejected = json['isRejected'];

        if (json['rejectedAt'] == null) {
            this.rejectedAt = null;
        } else if (typeof json['rejectedAt'] === 'string') {
            this.rejectedAt = parseISO(json['rejectedAt']);
        } else {
            throw new Error("Expected a datetime string 'rejectedAt'")
        }

        if (json['rejectedById'] == null) {
            this.rejectedById = null;
        } else if (isUUID(json['rejectedById'])) {
            this.rejectedById = json['rejectedById']
        } else {
            throw new Error("Expected a datetime string 'rejectedAt'")
        }

        if (!Array.isArray(json['allRejections']) || !json['allRejections'].every(isJsonObject)) {
            throw new Error("Expected a list of json objects 'allRejections'");
        }
        this.allRejections = json['allRejections'].map(allocationEventFromJson);

        if (typeof json['isDenied'] !== 'boolean') {
            throw new Error("Expected a boolean 'isDenied'");
        }
        this.isDenied = json['isDenied'];

        if (json['deniedAt'] == null) {
            this.deniedAt = null;
        } else if (typeof json['deniedAt'] === 'string') {
            this.deniedAt = parseISO(json['deniedAt']);
        } else {
            throw new Error("Expected a datetime string 'deniedAt'")
        }

        if (json['deniedById'] == null) {
            this.deniedById = null;
        } else if (isUUID(json['deniedById'])) {
            this.deniedById = json['deniedById']
        } else {
            throw new Error("Expected a datetime string 'deniedAt'")
        }

        if (json['setupBeginAt'] == null) {
            this.setupBeginAt = null;
        } else if (typeof json['setupBeginAt'] === 'string') {
            this.setupBeginAt = parseISO(json['setupBeginAt']);
        } else {
            throw new Error("Expected a datetime string 'setupBeginAt'")
        }

        if (json['setupById'] == null) {
            this.setupById = null;
        } else if (isUUID(json['setupById'])) {
            this.setupById = json['setupById']
        } else {
            throw new Error("Expected a datetime string 'setupAt'")
        }

        if (typeof json['isPrepared'] !== 'boolean') {
            throw new Error("Expected a boolean 'isPrepared'");
        }
        this.isPrepared = json['isPrepared'];


        if (json['preparedAt'] == null) {
            this.preparedAt = null;
        } else if (typeof json['preparedAt'] === 'string') {
            this.preparedAt = parseISO(json['preparedAt']);
        } else {
            throw new Error("Expected a datetime string 'preparedAt'")
        }

        if (typeof json['isCommenced'] !== 'boolean') {
            throw new Error("Expected a boolean 'isCommenced'");
        }
        this.isCommenced = json['isCommenced'];


        if (json['commencedAt'] == null) {
            this.commencedAt = null;
        } else if (typeof json['commencedAt'] === 'string') {
            this.commencedAt = parseISO(json['commencedAt']);
        } else {
            throw new Error("Expected a datetime string 'commencedAt'")
        }

        if (json['commencedById'] == null) {
            this.commencedById = null;
        } else if (isUUID(json['commencedById'])) {
            this.commencedById = json['commencedById']
        } else {
            throw new Error("Expected a datetime string 'commencedAt'")
        }

        if (!Array.isArray(json['progressEvents']) || !json['progressEvents'].every(isJsonObject)) {
            throw new Error("Expected a list of json objects 'progressEvents'");
        }
        this.progressEvents = json['progressEvents'].map(allocationEventFromJson)

        if (typeof json['isCompleted'] !== 'boolean') {
            throw new Error("Expected a boolean 'isCompleted'");
        }
        this.isCompleted = json['isCompleted'];

        if (json['completedAt'] == null) {
            this.completedAt = null;
        } else if (typeof json['completedAt'] === 'string') {
            this.completedAt = parseISO(json['completedAt']);
        } else {
            throw new Error("Expected a datetime string 'completedAt'")
        }

        if (json['completedById'] == null) {
            this.completedById = null;
        } else if (isUUID(json['completedById'])) {
            this.completedById = json['completedById']
        } else {
            throw new Error("Expected a datetime string 'completedAt'")
        }

        if (typeof json['isCancelled'] !== 'boolean') {
            throw new Error("Expected a boolean 'isCancelled'");
        }
        this.isCancelled = json['isCancelled'];


        if (json['cancelledAt'] == null) {
            this.cancelledAt = null;
        } else if (typeof json['cancelledAt'] === 'string') {
            this.cancelledAt = parseISO(json['cancelledAt']);
        } else {
            throw new Error("Expected a datetime string 'cancelledAt'")
        }

        if (json['cancelledById'] == null) {
            this.cancelledById = null;
        } else if (isUUID(json['cancelledById'])) {
            this.cancelledById = json['cancelledById']
        } else {
            throw new Error("Expected a datetime string 'cancelledAt'")
        }

        if (json['teardownBeginAt'] == null) {
            this.teardownBeginAt = null;
        } else if (typeof json['teardownBeginAt'] === 'string') {
            this.teardownBeginAt = parseISO(json['teardownBeginAt']);
        } else {
            throw new Error("Expected a datetime string 'teardownBeginAt'")
        }

        if (json['teardownById'] == null) {
            this.teardownById = null;
        } else if (isUUID(json['teardownById'])) {
            this.teardownById = json['teardownById']
        } else {
            throw new Error("Expected a datetime string 'teardownAt'")
        }

        if (typeof json['isFinalised'] !== 'boolean') {
            throw new Error("Expected a boolean 'isFinalised'");
        }
        this.isFinalised = json['isFinalised'];

        if (json['finalisedAt'] == null) {
            this.finalisedAt = null;
        } else if (typeof json['finalisedAt'] === 'string') {
            this.finalisedAt = parseISO(json['finalisedAt']);
        } else {
            throw new Error("Expected a datetime string 'finalisedAt'")
        }

        if (json['finalisedById'] == null) {
            this.finalisedById = null;
        } else if (isUUID(json['finalisedById'])) {
            this.finalisedById = json['finalisedById']
        } else {
            throw new Error("Expected a datetime string 'finalisedAt'")
        }
    }
}

export function labAllocationCreateForm() {
    const fb = inject(FormBuilder);
    return fb.group({
        startDate: fb.control<Date | null>(null),
        endDate: fb.control<Date | null>(null)
    });
}

export interface LabAllocationCreateRequest<TAllocation extends LabAllocation<any>> extends ModelCreateRequest<TAllocation> {
    consumerType: string;
    consumer: string;
    startDate: Date | null;
    endDate: Date | null;
}

export function labAllocationCreateRequestToJsonObject(request: LabAllocationCreateRequest<any>) {
    return {
        consumer: request.consumer,
        startDate: request.startDate ? formatISO(request.startDate) : null,
        endDate: request.endDate ? formatISO(request.endDate) : null,
    };
}
